service/evaluation/evaluation.go

```diff
@@ -444,80 +434,200 @@ func (svc *Service) addJobToScheduler(c *orchestrator.Control, toe *orchestrator
+			ValidManualOnly: util.Ref(true),
 		},
 		LatestByControlId: util.Ref(true),
-	})
+	},
+		func(ctx context.Context, req *evaluation.ListEvaluationResultsRequest, opts ...grpc.CallOption) (*evaluation.ListEvaluationResultsResponse, error) {
+			return svc.ListEvaluationResults(ctx, req)
+		}, func(res *evaluation.ListEvaluationResultsResponse) []*evaluation.EvaluationResult {
+			return res.Results
+		})
 	if err != nil {
-		err = fmt.Errorf("error list evaluation results: %w", err)
 		log.Error(err)
-		return
+		return err
+	}
+
+	manual = make(map[string][]*evaluation.EvaluationResult)
+
+	// Gather a list of controls, we are ignoring
+	ignored = make([]string, 0, len(results))
+	for _, result := range results {
+		if result.ParentControlId != nil {
+			manual[*result.ParentControlId] = append(manual[*result.ParentControlId], result)
+		} else {
+			ignored = append(ignored, result.ControlId)
+		}
+	}
+
+	// Filter relevant controls
+	for _, c := range controls {
+		// Only parent controls
+		if c.ParentControlId != nil {
+			continue
+		}
+
+		if c.IsRelevantFor(toe, catalog) {
+			relevant = append(relevant, c)
+		}
+
+		// If we ignore the control, we can skip it
+		if slices.Contains(ignored, c.Id) {
+			continue
+		}
 	}
 
-	// If no evaluation results for the sub-controls are available return and do not create a new evaluation result
-	if len(evaluations.Results) == 0 {
+	log.Infof("Starting catalog evaluation for Cloud Service '%s', Catalog ID '%s'. Waiting for the evaluation of %d control(s)",
+		toe.CloudServiceId,
+		toe.CatalogId,
+		len(relevant),
+	)
+
+	// We are using a timeout of half the interval, so that we are not running into overlapping executions
+	ctx, cancel = context.WithTimeout(ctx, time.Duration(interval)*time.Minute/2)
+	defer cancel()
+
+	g, ctx = errgroup.WithContext(ctx)
+	for _, control := range relevant {
+		control := control // https://golang.org/doc/faq#closures_and_goroutines needed until Go 1.22 (loopvar)
+		g.Go(func() error {
+			err := svc.evaluateControl(ctx, toe, catalog, control, manual[control.Id])
+			if err != nil {
+				return err
+			}
+
+			return nil
+		})
+	}
+
+	// Wait until all sub-controls are evaluated
+	err = g.Wait()
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	return nil
+}
+
+// evaluateControl evaluates a control, e.g., OPS-13. Therefore, the method needs to wait till all sub-controls (e.g.,
+// OPS-13.1) are evaluated.
+func (svc *Service) evaluateControl(ctx context.Context, toe *orchestrator.TargetOfEvaluation, catalog *orchestrator.Catalog, control *orchestrator.Control, manual []*evaluation.EvaluationResult) (err error) {
+	var (
+		status   = evaluation.EvaluationStatus_EVALUATION_STATUS_PENDING
+		result   *evaluation.EvaluationResult
+		results  []*evaluation.EvaluationResult
+		relevant []*orchestrator.Control
+		ignored  []string
+		g        *errgroup.Group
+	)
+
+	// Gather a list of sub control IDs that we have manual results for and thus we are ignoring
```

Comment:
Is it possible to outsource this to a new function? It is duplicated code with few exceptions which is also used in evaluateCatalog().

Reply:
I will try to do that in a future iteration